lib/collection/src/shards/transfer/shard_transfer.rs

```diff
@@ -246,31 +248,85 @@ async fn transfer_snapshot(
             ))
         })?;
 
-    // TODO: before transforing queue proxy and sending updates, make sure the remote shard is
-    // partial through consensus on all nodes
+    // TODO: through consensus, set replica state to partial
+
+    // TODO: await other node to have reached partial state
 
-    // Transfer all updates to remote shard and transform into forward proxy shard
-    // We do this right after the shard has been restored on the remote so that we can catch any
-    // errors early. The forward proxy shard we end up with will not error again once we unproxify.
     {
         let shard = shard_holder_read.get_shard(&shard_id);
-        if let Some(replica_set) = shard {
-            replica_set.queue_proxy_into_forward_proxy().await?;
-        } else {
+        let Some(replica_set) = shard else {
             return Err(CollectionError::service_error(format!(
                 "Shard {shard_id} cannot be transformed from queue to forward proxy because it does not exist"
             )));
-        }
+        };
+
+        // Transfer all updates to remote shard and transform into forward proxy shard
+        // We do this right after the shard has been restored on the remote so that we can catch any
+        // errors early. The forward proxy shard we end up with will not error again once we unproxify.
+        log::trace!("Transfer all queue proxy updates and transform into forward proxy");
+        replica_set.queue_proxy_into_forward_proxy().await?;
+
+        // Wait for remote shard to reach partial state in our replica set
+        log::trace!("Wait for local shard to reach Partial state");
+        replica_set
+            .wait_for_partial(transfer_config.to, defaults::CONSENSUS_META_OP_WAIT)
+            .await
+            .map_err(|err| {
+                CollectionError::service_error(format!(
+                    "Shard being transferred did not reach Partial state in time: {err}"
+                ))
+            })?;
     }
 
-    // We must keep partial state for 10 seconds to allow all nodes to catch up
-    // Future improvement: instead of waiting 10 seconds, confirm all nodes reached consensus
-    log::trace!("Shard snapshot transfer is waiting 10 seconds for consensus to catch up");
-    sleep(Duration::from_secs(10)).await;
+    // Synchronize, make sure all nodes have reached consensus before continuing and unproxying
+    // All other nodes must consider the shard state to be at least partial so that all nodes send
+    // operations to it.
+    await_consensus_sync(consensus, &channel_service, transfer_config.from).await;
```

Comment:
The essential integration.

- wait for shard to be in `Partial` state
- synchronize consensus across all nodes
- once all nodes reach consensus, complete transfer (and set shard to `Active` outside this function)